app/src/main/java/com/youssef/fixit/Auth/LoginFragment/LoginPresenter.java

```diff
@@ -0,0 +1,108 @@
+package com.youssef.fixit.Auth.LoginFragment;
+
+import android.content.Intent;
+import android.util.Log;
+import android.widget.Toast;
+
+import com.youssef.fixit.MainActivity.MainActivity;
+import com.youssef.fixit.Models.Bids.CreateBid;
+import com.youssef.fixit.Models.Data.RetrofitClient;
+import com.youssef.fixit.Models.Register.Register;
+
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+class LoginPresenter {
+    LoginView view;
+
+    public LoginPresenter(LoginView view) {
+        this.view = view;
+    }
+
+    void OnLogin(String Mail, String Password) {
+        Validation(Mail, Password);
+    }
+
+    void Validation(String Mail, String Password) {
+        if (Mail.isEmpty()) {
+            view.OnMailIsError("Please enter your Mail");
+            return;
+        }
+        if (Password.isEmpty()) {
+            view.OnPasswordIsError("Please enter your Password");
+            return;
+        }
+
+        LoginRequest(Mail, Password);
+    }
+
+    void LoginRequest(String Mail, String Password) {
+        RetrofitClient.getInstance().Login(Mail, Password).enqueue(new Callback<Register>() {
+            @Override
+            public void onResponse(Call<Register> call, Response<Register> response) {
+                if (response.isSuccessful() && response.code() == 200) {
+                    if (response.body().getData() != null) {
+                        List<String> roles = response.body().getData().getRoles();
```

Comment:
You can have a variable that holds the response data instead of accessing it everytime 

```suggestion
                       ... result = response.body().getData();
                       if(result !=null) {
                             List<String> roles = result.getRoles();
                             .....
                       }
```

What do you think, is it better? no?

Reply:
Yes, it is better